cli/cli_progress.go

```diff
@@ -3,127 +3,142 @@ package cli
 import (
 	"fmt"
 	"strings"
-	"sync"
+	"sync/atomic"
 	"time"
 
 	"github.com/kopia/kopia/internal/units"
-	"github.com/kopia/kopia/snapshot"
+	"github.com/kopia/kopia/snapshot/snapshotfs"
 )
 
-const maxUnshortenedPath = 60
+const spinner = `|/-\`
 
-type singleProgress struct {
-	desc      string
-	startTime time.Time
-	progress  int64
-	total     int64
+type cliProgress struct {
+	snapshotfs.NullUploadProgress
+
+	// all int64 must precede all int32 due to alignment requirements on ARM
+	uploadedBytes          int64
+	cachedBytes            int64
+	hashedBytes            int64
+	nextOutputTimeUnixNano int64
+
+	cachedFiles   int32
+	hashedFiles   int32
+	uploadedFiles int32
+
+	uploading      int32
+	uploadFinished int32
+
+	lastLineLength  int
+	spinPhase       int
+	uploadStartTime time.Time
 }
 
-func (p *singleProgress) update(progress, total int64) {
-	p.total = total
-	p.progress = progress
+func (p *cliProgress) FinishedHashingFile(fname string, totalSize int64) {
+	atomic.AddInt32(&p.hashedFiles, 1)
+	p.maybeOutput()
 }
 
-func (p *singleProgress) toString(details bool) string {
-	if p.total == 0 {
-		return fmt.Sprintf("empty %v", p.desc)
-	}
+func (p *cliProgress) UploadedBytes(numBytes int64) {
+	atomic.AddInt64(&p.uploadedBytes, numBytes)
+	atomic.AddInt32(&p.uploadedFiles, 1)
```

Comment:
Does a call here really correspond to a (source) file being uploaded? or does it (strictly) refer to uploaded blobs?
In case it is the latter, it makes sense to rename the field, maybe to `uploadedBlobs` ?
In either case, it'd be good to expand the documentation for `UploadBytes` in the interface for clarity.